### PR TITLE
[Unification] Generalize TransformerEncoder

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #253
* #249
* #248
* #244
* #243
* __->__ #240
* #239

Differential Revision: [D38506881](https://our.internmc.facebook.com/intern/diff/D38506881)

## Summary
Add a general `TransformerEncoder` class that simply stacks `n` layers of our custom `TransformerEncoderLayer`. Repurpose `FLAVATransformerOutput` and use it as `TransformerOutput` for this class.

## Test plan
Newly added unit tests, `pytest test/modules/layers/test_transformer.py -vv`
```
===================================================== test session starts ======================================================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0 -- /fsx/users/rafiayub/conda/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /data/home/rafiayub/torchmultimodal, configfile: pyproject.toml
plugins: hydra-core-1.1.2, cov-3.0.0, mock-3.8.2
collected 5 items                                                                                                              

test/modules/layers/test_transformer.py::TestTransformerEncoderLayer::test_forward_prenorm PASSED                        [ 20%]
test/modules/layers/test_transformer.py::TestTransformerEncoderLayer::test_forward_postnorm PASSED                       [ 40%]
test/modules/layers/test_transformer.py::TestTransformerCrossAttentionLayer::test_forward_prenorm PASSED                 [ 60%]
test/modules/layers/test_transformer.py::TestTransformerCrossAttentionLayer::test_forward_postnorm PASSED                [ 80%]
test/modules/layers/test_transformer.py::TestTransformerEncoder::test_forward PASSED                                     [100%]

====================================================== 5 passed in 1.59s =======================================================
```

### CODE REVIEW

File: torchmultimodal/modules/layers/transformer.py [hunk 1]
```diff
@@ -7,12 +7,27 @@
...
 
-import torch
 from torch import nn, Tensor
 from torchmultimodal.modules.layers.attention import MultiHeadAttention, SelfAttention
 from torchmultimodal.modules.layers.mlp import MLP
+from torchmultimodal.modules.layers.normalizations import fp32layernorm
+from torchmultimodal.utils.common import get_clones
+
+
+TransformerOutput = namedtuple(
```
Comment (langong347) [inline torchmultimodal/modules/layers/transformer.py:20]: Should we just uniformly subclass from `NamedTuple` for model outputs?

File: torchmultimodal/modules/layers/transformer.py [hunk 2]
```diff
@@ -328,11 +341,71 @@ def forward(
...
+        return_hidden_states: bool = False,
+    ) -> TransformerOutput:
+        all_hidden_states: Tuple[Tensor, ...] = () if return_hidden_states else None
+        all_self_attentions: Tuple[Tensor, ...] = () if return_attn_weights else None
+
+        for i, layer_module in enumerate(self.layer):
+            if return_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+
+            layer_head_mask = head_mask[i] if head_mask is not None else None
```
Comment (langong347) [inline torchmultimodal/modules/layers/transformer.py:369]: head_masks can be different across the layers?

File: torchmultimodal/modules/layers/transformer.py [hunk 3]
```diff
@@ -328,11 +341,71 @@ def forward(
...
+            layer_outputs = layer_module(
+                hidden_states,
+                attention_mask=attention_mask,
+                head_mask=layer_head_mask,
+                return_attn_weights=True,
+            )
+
+            hidden_states = layer_outputs[0]
```
Comment (langong347) [inline torchmultimodal/modules/layers/transformer.py:377]: If `layer_outputs` only contains the hidden_states, this will try to index the 0-th element instead of returning the full hidden_states, right?

```
        if return_attn_weights:
            return outputs, attn_weights
        else:
            return outputs
```

Consider adding a unit test when the kwargs args are False as well?

```
output = encoder(inputs, return_hidden_states=True, return_attn_weights=True)
```

File: torchmultimodal/modules/layers/normalizations.py [hunk 4]
```diff
@@ -22,3 +23,13 @@ def forward(self, x: Tensor) -> Tensor:
             self.eps,
         )
         return output.type_as(x)
+
+
+def fp32layernorm(x: Tensor, layernorm: nn.Module) -> Tensor:
```
Comment (langong347) [inline torchmultimodal/modules/layers/normalizations.py:28]: Why need the functional besides `class Fp32LayerNorm`?

File: torchmultimodal/modules/layers/transformer.py (hunk 3 above)
Comment (langong347) [reply torchmultimodal/modules/layers/transformer.py]: Not sure if this has been addressed as the encoder layer doesn't always return two tensors, right?

https://github.com/facebookresearch/multimodal/blob/adc36c18c0ccae7652eee34256ecc74e628f7aba/torchmultimodal/modules/layers/transformer.py#L381

Comment (RdoubleA) [reply torchmultimodal/modules/layers/transformer.py]: return_attn_weights is set to True so it will guarantee returning two outputs

Comment (langong347) [reply torchmultimodal/modules/layers/transformer.py]: Not sure. Then we should probably ask `TransformerEncoderLayer` to always return two outputs anyways.

Comment (RdoubleA) [reply torchmultimodal/modules/layers/transformer.py]: hmm, maybe I'll modify to handle both single and double outputs

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506881).

**Review by langong347 — APPROVED** [review]

File: torchmultimodal/modules/layers/transformer.py [hunk 5]
```diff
@@ -328,11 +334,68 @@ def forward(
...
+                    dim_feedforward,
+                    dropout,
+                    activation,
+                    layer_norm_eps,
+                    norm_first,
+                )
+                for _ in range(n_layer)
+            ]
+        )
+        self.num_layers = n_layer
```
Comment (ankitade) [inline torchmultimodal/modules/layers/transformer.py:364]: not used?

File: torchmultimodal/modules/layers/transformer.py [hunk 6]
```diff
@@ -328,11 +334,68 @@ def forward(
...
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerOutput:
+        all_hidden_states: Tuple[Tensor, ...] = () if return_hidden_states else None
+        all_self_attentions: Tuple[Tensor, ...] = () if return_attn_weights else None
+
+        for i, layer_module in enumerate(self.layer):
```
Comment (ankitade) [inline torchmultimodal/modules/layers/transformer.py:377]: you dont need enumeration

File: torchmultimodal/modules/layers/transformer.py [hunk 7]
```diff
@@ -328,11 +334,68 @@ def forward(
...
+            if return_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
```
Comment (ankitade) [inline torchmultimodal/modules/layers/transformer.py:379]: hm will this add the original input hidden states too?

Comment (RdoubleA) [reply torchmultimodal/modules/layers/transformer.py:379]: this is straight from the original `FLAVATransformerEncoder`: https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/modules/layers/transformer.py#L384

File: torchmultimodal/modules/layers/transformer.py (hunk 5 above)
Comment (RdoubleA) [reply torchmultimodal/modules/layers/transformer.py:364]: hmmm I think it was used in [CLIPTextEncoder](https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/modules/encoders/clip_text_encoder.py#L80). I won't be able to get to unifying that component but maybe we should keep `self.num_layers` for when we do?

Comment (ankitade) [reply torchmultimodal/modules/layers/transformer.py:364]: i think the caller can maintain the num layers. no need to add unused instance variables here

Comment (RdoubleA) [reply torchmultimodal/modules/layers/transformer.py:364]: aight I'll remove it

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506881).

**Review by ankitade — APPROVED** [review]

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506881).
